adal/src/telemetry/java/com/microsoft/aad/adal/APIEvent.java

```diff
@@ -0,0 +1,172 @@
+ * This class tracks flows for certain API calls. Most notably in those all acquireToken* calls
+ * All other Events will be called within the timeline of APIEvent
+ */
+class APIEvent extends DefaultEvent {
+
+    private static final String TAG = DefaultEvent.class.getSimpleName();
+    private final String mEventName;
+
+    APIEvent(final String eventName) {
+        setEvent(EventStrings.EVENT_NAME, eventName);
+        mEventName = eventName;
+    }
+
+    APIEvent(final String eventName, final Context context, final String clientId) {
+        this(eventName);
+        setDefaults(context, clientId);
+    }
+
+    void setAuthority(final String authority) {
+        if (StringExtensions.isNullOrBlank(authority)) {
+            return;
+        }
+
+        setEvent(EventStrings.AUTHORITY_NAME, authority);
+        final URL authorityUrl = StringExtensions.getUrl(authority);
+        if (authorityUrl == null) {
+            return;
+        }
+
+        if (UrlExtensions.isADFSAuthority(authorityUrl)) {
+            setAuthorityType(EventStrings.AUTHORITY_TYPE_ADFS);
+        } else {
+            setAuthorityType(EventStrings.AUTHORITY_TYPE_AAD);
+        }
+    }
+
+    void setAuthorityType(final String authorityType) {
+        setEvent(EventStrings.AUTHORITY_TYPE, authorityType);
+    }
+
+    void setIsDeprecated(final Boolean isDeprecated) {
+        setEvent(EventStrings.API_DEPRECATED, isDeprecated.toString());
+    }
+
+    void setValidationStatus(final String validationStatus) {
+        setEvent(EventStrings.AUTHORITY_VALIDATION, validationStatus);
+    }
+
+    void setExtendedExpiresOnSetting(final Boolean extendedExpiresOnSetting) {
+        setEvent(EventStrings.EXTENDED_EXPIRES_ON_SETTING, extendedExpiresOnSetting.toString());
+    }
+
+    void setSilentRequestPromptBehavior(final String promptBehavior) {
+        setEvent(EventStrings.PROMPT_BEHAVIOR, promptBehavior);
+    }
+
+    void setAPIId(final String id) {
+        setEvent(EventStrings.API_ID, id);
+    }
+
+    String getEventName() {
+        return mEventName;
+    }
+
+    void setWasApiCallSuccessful(final Boolean isSuccess, final Exception exception) {
+        setEvent(EventStrings.WAS_SUCCESSFUL, isSuccess.toString());
+
+        if (exception != null) {
+            if (exception instanceof AuthenticationException) {
+                final AuthenticationException authException = (AuthenticationException) exception;
+                setEvent(EventStrings.API_ERROR_CODE, authException.getCode().toString());
+            }
+        }
+    }
+
+    void stopTelemetryAndFlush() {
+        Telemetry.getInstance().stopEvent(getRequestId(), this, getEventName());
+        Telemetry.getInstance().flush(getRequestId());
+    }
+
+    void setIdToken(final String rawIdToken) {
+        if (StringExtensions.isNullOrBlank(rawIdToken)) {
+            return;
+        }
+
+        final IdToken idToken;
+        try {
+            idToken = new IdToken(rawIdToken);
```

Comment:
why is the ID token being parsed again? all this information should be in the authentication result.